x/evm/keeper/grpc_query.go

```diff
@@ -393,3 +400,100 @@ func (k Keeper) EthCall(c context.Context, req *types.EthCallRequest) (*types.Ms
 
 	return res, nil
 }
+
+// EstimateGas implements eth_estimateGas rpc api.
+func (k Keeper) EstimateGas(c context.Context, req *types.EthCallRequest) (*types.EstimateGasResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	k.WithContext(ctx)
+
+	var args types.CallArgs
+	err := json.Unmarshal(req.Args, &args)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// Binary search the gas requirement, as it may be higher than the amount used
```

Comment:
it's interesting that the binary search on geth is used by the public api side meaning it is done by the client of this grpc server.

At the moment the only relevant issue I've found is https://github.com/ethereum/go-ethereum/issues/21769

I think we should do the binary search on the public API (grpc client side) for sake of matching what they do in Geth

> one rpc call emites several grpc query logs.

But this is a reasonable pattern right? I think we can adjust the logs to make it more clear that these are part of the same process... effectively it is slow and painful to estimate gas so logging completions per call is not necessarily bad

@fedekunze also any thoughts on making these RPC server files `grpc_query.go` files `grpc_server.go` - you might have more context behind the naming but I think it's less obvious.

Reply:
One thing that's less clean when doing binary search in rpc api side(or grpc client side) is checking the error type, in grpc server side, you can inspect the error type directly, in grpc client side, you have to check a sub string in the error string which might not be accurate.

Reply:
This has to do with not being able to use `error.Is` right? Are you sure that's the case? `doCall` for example will run through a contract call and expose these error types through `vmError` in Geth, we can do this as well?

Reply:
> This has to do with not being able to use `error.Is` right? Are you sure that's the case? `doCall` for example will run through a contract call and expose these error types through `vmError` in Geth, we can do this as well?

Do you mean `MsgEthereumTxResponse.VmError`? that is serialized version of [these errors](https://github.com/ethereum/go-ethereum/blob/master/core/vm/errors.go#L26), when these errors are not further wrapped, you can compare the error string as a whole. But if the error is [wrapped with stacktrace](https://github.com/tharsis/ethermint/blob/main/x/evm/keeper/state_transition.go#L249) and [grpc sdk errors](https://github.com/tharsis/ethermint/blob/main/x/evm/keeper/grpc_query.go#L391), the best you can do in the grpc client side is `strings.Contains`.